chore!: remove backend interactions from `nargo`

TomAFrench opened the pull request:
This PR removes the `nargo codegen-verifier` command and replaces it with an example bash script showing how to use `nargo` in conjunction with `bb` to generate a smart contract verifier.

cc @AztecProtocol/devrel 



Closes https://github.com/noir-lang/noir/issues/4447
Closes https://github.com/noir-lang/noir/pull/1138
Closes https://github.com/noir-lang/noir/issues/2222
Closes https://github.com/noir-lang/noir/issues/1159
Closes https://github.com/noir-lang/noir/issues/1009
Closes https://github.com/noir-lang/noir/issues/3837
Closes https://github.com/noir-lang/noir/issues/2667

TomAFrench committed:
chore: add example script to generate a verifier contract

TomAFrench committed:
chore!: remove `codegen-verifier` command

signorecello reviewed (dismissed):
Reached out on slack, this will make 90% of Noir devs install a new package (+node if they don't have it already). Let's see if we can find a middle ground solution 👍

TomAFrench commented:
As discussed on slack, the solution is for us to have good interop between nargo artifacts and `bb` as done in #6283, we should also expect that users should install their proving backend of choice rather than nargo acting as an installer for `bb`.

TomAFrench commented:
Note that the UX from this change goes from

```
nargocodegen-verifier
```
to

```
nargo compile
bb write_vk -b ./target/hello_world.json
bb contract -v ./target/vk
```

which after some UX changes to `bb` could become

```
nargo compile
bb contract
```

Savio-Sou commented:
> which after some UX changes to bb could become
> ```
> nargo compile
> bb contract
> ```

This `nargo codegen-verifier` PR should aim for this as the goal rather than something in between.

A requirement to https://github.com/noir-lang/noir/issues/4960 is we do our best attempts at minimizing additional DevEx frictions coming out of it. If that means more engineering time, we allocate the time 🤝

TomAFrench commented:
Happy to discuss audit timelines if we're taking Noir development resources to rewrite all of Noir's workspace logic in C++.

Savio-Sou commented:
Good point.

To clarify I was thinking of an aim to add default behaviors (if they don't already exist):
1. Update `bb write_vk` (flagless) to operate on `./target/main.json` by default
2. Update `bb contract` (flagless) to operate on `./target/vk` by default
3. Update `bb contract` (flagless) to call `bb write_vk` (flagless) automatically if `./target/vk` doesn't already exist
While maintaining flag options for users that need them to operate on multiple / renamed files.

I don't think there are a lot of user demand for:
- Generate multiple Solidity verifiers
- For multiple Noir programs
- Within one single workspace
at least today (@AztecProtocol/devrel to confirm).

On that assumption, offering a reference bash script and kicking the `bb` workspace can down the road is acceptable.

signorecello commented:
I do lack full context on the end goal here, but I understand there is a significant engineering reason for burning this nargo-bb bridge. So I won't say we shouldn't do it, but I do care about the devex since that's the biggest Noir strength.

For context, every single Noir workshop we do starts with generating a proof and a smart contract, and verifying it on Remix IDE. Literally every one of the 60+ Noir submissions at the Scroll hackathon generated a smart contract with Nargo.

Downgrading the devex introducing more packages and commands to something so ubiquitous seems like something we want to avoid, if possible.

I proposed a solution, which I reckon is not ideal: to have DevRel maintain a CLI tool that that could encapsulate Nargo and BB, essentially delivering the same execution + proving devex. I told @critesjosh I could take on that task, but I think it's worth exploring if there's any other way

TomAFrench committed:
chore: update docs

TomAFrench commented on noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh:
```diff
@@ -0,0 +1,10 @@
+
+# TODO: backend should automatically generate vk if necessary.
+$BACKEND write_vk -b ./target/hello_world.json
+$BACKEND contract -v ./target/vk
```
```suggestion
$BACKEND contract
```

TomAFrench committed:
Update noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh

TomAFrench committed:
chore: update integration tests to call `bb` directly

Savio-Sou commented:
@signorecello we surely don't want to build and maintain a new wrapper. That would undermine the whole point of https://github.com/noir-lang/noir/issues/4960 for cleaner engineering scope (hence simpler auditing) and improved DevEd (hence DevEx).

## DevEx: aiming for net positive

The initiative aims at cleanly separating Noir and Backend tooling **from a user perspective**, which helps familiarize users with the Noir <> Backend concept as clear and as soon during their onboarding as possible. This in turn helps users to more accurately debug issues they run into, as well as paves way for users to more easily work between different backends.

> there is a significant engineering reason for burning this nargo-bb bridge

This hence is a tradeoff between different verticals of DevEx, rather than a net DevEx negative in exchange for some obscure internal engineering goals.

## DevEx Change: Single Solidity Verifier

To clarify on effects, with my suggestion above the DevEx for generating **one** Solidity verifier would change from:

```
nargo codegen-verifier
```

to:

```
nargo compile
bb contract
```

We would already be at the ceiling of intuitiveness on separating the two toolings.

## DevEx Change: Multiple Solidity Verifier

The intermediate DevEx deterioration (until workspace support is introduced in barretenberg tooling) lies with generating **multiple** Solidity verifiers, which would change from:

```
nargo codegen-verifier
```

to:

```
(create / edit bash script)
bash codegen_verifier.sh
```

This was assumed acceptable as little to none user demand was assumed for frequently generating multiple Solidity verifiers. Do comment below if that isn't accurate 🙏

TomAFrench committed:
chore: update usage to handle bb.js

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

TomAFrench committed:
chore: add simple docs for installing `bb`

signorecello commented:
> ## DevEx Change: Single Solidity Verifier
> To clarify on effects, with my suggestion above the DevEx for generating **one** Solidity verifier would change from:
> 
> ```
> nargo codegen-verifier
> ```
> 
> to:
> 
> ```
> nargo compile
> bb contract
> ```
> 
> We would already be at the ceiling of intuitiveness on separating the two toolings.

I agree we mostly care about 80% of users, who would be generating one Solidity verifier only. 

But the proposed change is not as easy as just using two commands instead of one. Some problems come to mind:
1. `noirup` installs `nargo` and according to Tom it won't install `bb`. Will there be a similar, easy script for `bb`? `bbup`? And in that case, will there be a script that runs both `noirup` and `bbup`? Doesn't that essentially mean the proposal above that, I agree, will take too much effort?
2. Nargo documentation lives in Noir docs. Where will the documentation for `bb` live? I talked with @codygunton some time ago on practicing a bit with the Doxygen config they have, but decided to spare them from our incessant requests and docusaurus stuff.
3. Related to the above, who exactly will own BB? If the solidity verifier breaks, I usually just reach out to the Noir team... Burning this bridge, we risk putting BB even more on the outside. There's no BB team.

I don't think the proposed separation on itself would mean net positive DevEx. Net positive DevEd, yes, as the user understands the clear separation between compiler and prover. But strictly speaking, this means more packages and more commands. That's not net positive DevEx. Even if we have an all-in-on `noirbbup` package as suggested above, the user still needs to install two packages, ensure their versions are compatible, use two commands, probably looking at two different websites for their documentation, etc.

I'm just concerned about throwing away some of the biggest assets we have: the all-mighty Nargo. People are delighted when they see it and realize "wait this thing compiles, tests, executes, proves, generates smart contract, all in one?"

Savio-Sou commented:
@signorecello happy to talk through (1), (2) and (3), but maybe somewhere else to keep the thread focused on this PR.

> we mostly care about 80% of users

I was thinking more >95% of users aren't frequently generating multiple Solidity verifiers. Were you simply speaking idiomatically, or are there projects that you know of that relies heavily on it?

TomAFrench committed:
chore: fix codegen-verifiers.sh

TomAFrench commented:
> Will there be a similar, easy script for bb? bbup?

Yeah, you could create a `bbup` script from `noirup` pretty easily. Just need to delete all of the code related to building from source.

> Nargo documentation lives in Noir docs. Where will the documentation for bb live?

We should have examples of usage in the Noir docs for the time being. There should eventually be docs in the barretenberg repository for how to install/use `bb` at some point though imo, no need for a full docusaurus setup. Having decent output for `bb --help` would help greatly also.

TomAFrench committed:
chore: installs `bb` for noir CI

TomAFrench commented:
@signorecello I've added docs on how to install `bb` to this PR, updated the "how to" to show usage for generating verifier contracts. There's another example added to the repo which ensures that `nargo` and `bb` work together.

Some simple installation tooling for `bb` is proposed in #6376, `bb` will need to be documented as it's one of Aztec's user-facing products.

> Even if we have an all-in-on noirbbup package as suggested above, the user still needs to install two packages, ensure their versions are compatible, use two commands, probably looking at two different websites for their documentation, etc.

Yes, it is correct that we use two different commands as we're interacting with two different programs doing two different things. Trying to hide this fact results in us having to add a bunch of complication to force all inputs/outputs to be completely standardised as well as reducing flexibility (`bb` exposes several flavours of the proving system atm but `nargo prove` can only use one of them.

The problem of ACIR compatibility is a problem which will be experienced by all ACIR backends and this can't just be solved by nargo always going off and downloading the relevant version of whichever backend the user is using as this doesn't scale.

TomAFrench committed:
feat: extend example to show how

TomAFrench committed:
Update noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

TomAFrench committed:
pull over some more files

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

TomAFrench committed:
chore: add `examples` target to CI

TomAFrench committed:
chore: update test

TomAFrench committed:
chore: remove contract.sol

ludamad committed:
chore(ci): dont look for tags if pure spot (#6446)

This caused testers to look for invalid tags

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

vezenovm commented on noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh:
```diff
@@ -0,0 +1,44 @@
+HEX_PUBLIC_INPUTS=$(head -c $PUBLIC_INPUT_BYTES $PROOF_PATH | od -An -v -t x1 | tr -d $' \n')
+HEX_PROOF=$(tail -c +$(($PUBLIC_INPUT_BYTES + 1)) $PROOF_PATH | od -An -v -t x1 | tr -d $' \n')
+
+# echo "Public inputs:"
+# echo $HEX_PUBLIC_INPUTS
+
+# echo "Proof:"
+# echo "0x$HEX_PROOF"
+
```
```suggestion
```
Looks like leftover debug stuff?

vezenovm commented on noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh:
```diff
@@ -0,0 +1,44 @@
+
+nargo compile
+
+# TODO: backend should automatically generate vk if necessary.
```
Should we make an issue on barretenberg repo or do we have one?

vezenovm commented on noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh:
```diff
@@ -0,0 +1,44 @@
+VERIFIER_ADDRESS=$(echo $DEPLOY_INFO | jq -r '.deployedTo')
+
+# Call the verifier contract with our proof.
+# Note that we haven't needed to split up `HEX_PUBLIC_INPUTS` as there's only a single public input
```
Another TODO should maybe be to make it easier to get a proof with the public inputs split off from bb. A dev should not have to specify the `NUM_PUBLIC_INPUTS` as we do in this script as they can get it from the vkey. This  is non-blocking just noting for a follow-up

TomAFrench committed:
Update noir/noir-repo/examples/codegen-verifier/codegen_verifier.sh

Co-authored-by: Maxim Vezenov <mvezenov@gmail.com>

TomAFrench commented:
Opened issues.

TomAFrench committed:
chore!: remove `nargo prove` and `nargo verify` (#6321)

TomAFrench committed:
feat!: remove backend interactions from `nargo` (#6369)

This is the nuclear option of separating `nargo` and `bb`. Nargo no
longer downloads or interacts with `bb` at all.

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

TomAFrench commented on noir/noir-repo/docs/docs/how_to/how-to-solidity-verifier.md:
```diff
@@ -123,11 +131,25 @@ To verify a proof using the Solidity verifier contract, we call the `verify` fun
+NUM_PUBLIC_INPUTS=1
+PUBLIC_INPUT_BYTES=32*NUM_PUBLIC_INPUTS
+HEX_PUBLIC_INPUTS=$(head -c $PUBLIC_INPUT_BYTES ./proof | od -An -v -t x1 | tr -d $' \n')
+HEX_PROOF=$(tail -c +$PUBLIC_INPUT_BYTES ./proof | od -An -v -t x1 | tr -d $' \n')
```
```suggestion
HEX_PROOF=$(tail -c +$(($PUBLIC_INPUT_BYTES + 1)) ./proof | od -An -v -t x1 | tr -d $' \n')
```

TomAFrench committed:
Update noir/noir-repo/docs/docs/how_to/how-to-solidity-verifier.md

TomAFrench committed:
chore: remove dead code

TomAFrench commented:
Closes https://github.com/noir-lang/noir/issues/4447
Closes https://github.com/noir-lang/noir/pull/1138
Closes https://github.com/noir-lang/noir/issues/2222
Closes https://github.com/noir-lang/noir/issues/1159
Closes https://github.com/noir-lang/noir/issues/1009
Closes https://github.com/noir-lang/noir/issues/3837
Closes https://github.com/noir-lang/noir/issues/2667

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

TomAFrench committed:
Merge branch 'master' into tf/remove-codegen-verifier

TomAFrench committed:
fmt

TomAFrench committed:
chore: update docs to reflect lack of `Verifier.toml` file
